pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+          _clusterInfoAccessor).entrySet()) {
+        for (PinotTaskConfig taskConfig : _clusterInfoAccessor.getTaskConfigs(entry.getKey())) {
+          inCompleteMergeLevelToTaskConfig.computeIfAbsent(taskConfig.getConfigs().get(MergeRollupTask.MERGE_LEVEL_KEY),
+              k -> new ArrayList<>()).add(taskConfig);
+          segmentsScheduled.addAll(Arrays.asList(taskConfig.getConfigs()
+              .get(MinionConstants.SEGMENT_NAME_KEY)
+              .split(MinionConstants.SEGMENT_NAME_SEPARATOR)));
+        }
+      }
+
+      // Sort segments based on startTimeMs and endTimeMs in ascending order
+      TreeSet<OfflineSegmentZKMetadata> preSelectedSegmentsForOfflineTable = allSegmentsForOfflineTable.stream()
+          .filter(m -> selectedSegmentsBasedOnLineage.contains(m.getSegmentName()) && !segmentsScheduled.contains(
+              m.getSegmentName()))
+          .collect(Collectors.toCollection(() -> new TreeSet<>((a, b) -> {
+            long aStartTime = a.getStartTimeMs();
+            long bStartTime = b.getStartTimeMs();
+            return aStartTime != bStartTime ? Long.compare(aStartTime, bStartTime)
+                : Long.compare(a.getEndTimeMs(), b.getEndTimeMs());
+          })));
+
+      if (preSelectedSegmentsForOfflineTable.isEmpty()) {
+        LOGGER.warn("Skip generating task: {} for table: {}, no segment is found to merge.", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      // Sort merge levels based on bucket time period
+      Map<String, Map<String, String>> mergeLevelToConfigs = MergeRollupTaskUtils.getLevelToConfigMap(taskConfigs);
+      List<Map.Entry<String, Map<String, String>>> sortedMergeLevelConfigs = mergeLevelToConfigs.entrySet()
+          .stream()
+          .sorted(Comparator.comparingLong(
+              e -> TimeUtils.convertPeriodToMillis(e.getValue().get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY))))
+          .collect(Collectors.toList());
+
+      // Schedule tasks from lowest to highest merge level
+      String lowerMergeLevel = null;
+      long lowerMergeLevelWatermarkMs = -1;
+
+      for (Map.Entry<String, Map<String, String>> entry : sortedMergeLevelConfigs) {
+        String mergeLevel = entry.getKey();
+        Map<String, String> mergeConfig = entry.getValue();
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode.
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(offlineTableName, preSelectedSegmentsForOfflineTable.first().getStartTimeMs(), bucketMs,
```

Comment:
We should not read the watermark ZK metadata for each level. We can read it once, calculate and update the watermark for all merge levels, then persist it back once with the version check.

Reply:
Good point, updated the pr to cache the ZK metadata, and only update it per table with version check.